src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+                var text = location.SourceTree.GetText();
+                var textLines = text.Lines;
+
+                var lineSpan = location.GetLineSpan(); // FileLinePositionSpan type.
+                var errorLineNumber = lineSpan.StartLinePosition.Line;
+                var errorPosition = lineSpan.StartLinePosition.Character;
+
+                // This is typical Roslyn diagnostic message which contains
+                // a message number, a source context and an error position.
+                var diagnisticMessage = diagnisticRecord.ToString();
+                var errorLineString = textLines[errorLineNumber].ToString();
+
+                StringBuilder sb = new StringBuilder(diagnisticMessage.Length + errorLineString.Length * 2 + 4);
+
+                sb.AppendLine(diagnisticMessage);
+                sb.AppendLine(errorLineString);
+
+                for (var i = 0 ; i < errorLineString.Length ; i++)
                 {
-                    ErrorText = d.GetMessage(),
-                    FileName = null,
-                    Line = d.Location.GetMappedLineSpan().StartLinePosition.Line + 1, // Convert 0-based to 1-based
-                    IsWarning = !d.IsWarningAsError && d.Severity == DiagnosticSeverity.Warning,
-                    Column = d.Location.GetMappedLineSpan().StartLinePosition.Character + 1, // Convert 0-based to 1-based
-                    ErrorNumber = d.Id
-                }).ToArray();
+                    if (errorLineString[i] != ' ' && errorLineString[i] != '\t')
+                    {
+                        // We copy white chars from the source string.
+                        sb.Append(errorLineString, 0, i);
+                        // then pad up to the error position.
+                        sb.Append(' ', Math.Max(0, lineSpan.StartLinePosition.Character - i));
+                        // then put "^" into the error position.
+                        sb.AppendLine("^");
+                        break;
+                    }
+                }
+
+                return sb.ToString();
+            }
         }
+
+        private static int ArraySyntaxTreeGetHashCode(IEnumerable<SyntaxTree> sts)
+        {
+            // We use our extension method EnumerableExtensions.SequenceGetHashCode<T>().
+            List<int> stHashes = new List<int>();
+            foreach (var st in sts)
+            {
+                stHashes.Add(SyntaxTreeGetHashCode(st));
+            }
+
+            return stHashes.SequenceGetHashCode<int>();
+        }
+
+        private static int SyntaxTreeGetHashCode(SyntaxTree st)
+        {
+            int hash;
+
+            if (String.IsNullOrEmpty(st.FilePath))
+            {
+                // If the file name does not exist, the source text is set by the user using parameters.
+                // In this case, we assume that the source text is of a small size and we can re-allocate by ToString().
+                hash = st.GetText().ToString().GetHashCode();
```

Comment:
Based on the [comments](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Syntax/SyntaxTree.cs#L347) in `SyntaxTree`, you can directly call `ToString()` without having to call `GetText()` first.

Reply:
Fixed.